### ISSUE TITLE
Selected value not showing in the input field [Controlled component, single select]

### ISSUE THREAD

**Issue by stheurka** [JedWatson/react-select#2674]
Selected value does not show in the input field for a controlled react-select component with single select.

Demo:
https://codesandbox.io/s/ypwq481j7z

Code:
`import React, { Component} from "react";
import Select from "react-select";

type State = {
  selectedValue: string
};

export default class SingleSelect extends Component<*, State> {
  state = {
    selectedValue: ""
  };

  optionChanged = value => {
    console.log(value);
    this.setState({ selectedValue: value});
  };
  render() {
    console.log(this.state.selectedValue);
    return (
      <div>
        <Select
          options={["abcd", "def", "ghi"]}
          value={this.state.selectedValue}
          onChange={this.optionChanged}
          getOptionLabel={option => option}
          getOptionValue={option => option}
        />
      </div>
    );
  }
}
`

**Comment by sbrichardson** [comment]
Yeah, I'm not sure which issue to report in, but I can't get a simple value to display after migrating. It set's the value from an onChange correctly, and the props update and console.log properly, but the component doesn't visually show the new value.


> Here is the OP code formatted for assistance (not my code)

```js
import React, { Component } from 'react'
import Select from 'react-select'

type State = {
  selectedValue: string
}

export default class SingleSelect extends Component<*, State> {
  state = {
    selectedValue: ''
  }

  optionChanged = value => {
    console.log(value)
    this.setState({ selectedValue: value })
  }

  render() {
    console.log(this.state.selectedValue)
    return (
      <Select
        options={['abcd', 'def', 'ghi']}
        value={this.state.selectedValue}
        onChange={this.optionChanged}
        getOptionLabel={option => option}
        getOptionValue={option => option}
      />
    )
  }
}

```

**Comment by TheOld** [comment]
Yeah, I'm having the exact same issue. I even tried using one of the code example from the docs and not even those are working.

**Comment by JanBN** [comment]
same problem.

**Comment by avernikoz** [comment]
The same problem.

**Comment by avernikoz** [comment]
Guys, I found a workaround how to avoid it.
First of all, you should save your value in state **not in string type**, but in **array type**, even if you don't use multi-select!

`state: 
    changeInPercentage: []

handleOnchange = (optionSelected: {label: string, value: string}) => {
this.setState({
      changeInPercentage: [
        {
          label: optionSelected.label,
          value: value
        }
      ]
    }, () => {
      console.log(this.state);
    })
}

<SelectR
styles={customStyles}
isClearable
options={data.changeInPercentage}
onChange={this.handleSelectChangeWithInput.bind(this, 'changeInPercentage')}
value={this.state.changeInPercentage}
/>
`

It happens, because in the source-code of Select v2 there is a functions, that pass only array-typed values:
`
export const cleanValue = (value: ValueType): OptionsType => {
  if (Array.isArray(value)) return value.filter(Boolean);
  if (typeof value === 'object' && value !== null) return [value];
  return [];
};

`

**Comment by avernikoz** [comment]
@JedWatson Can I take a look at this function and how to reimplement it, or maybe you have something in your drafts?

**Comment by guillaumev** [comment]
Same issue here. The workaround works but it is a lot of code for a simple select component with a few elements in it...

**Comment by 27percent** [comment]
@jossmac that seems really counter-intuitive for a select menu, supplying a value that gets matched with one of the values in the options array to get the label seems like a more natural API. 
You could potentially support both though - if the value is of shape `Array<Object>` then use the existing API, if its a simple string, try to match it with one of the options. I'd be happy to do the work and put in a PR if the suggestion sounds sensible to you :)

**Comment by jossmac** [comment]
@27percent if you need simple value, consider using a helper function:

```jsx
const valueFromId = (opts, id) => opts.find(o => o.value === id);

<Select value={valueFromId(this.props.options, this.props.id) />
```

**Comment by 27percent** [comment]
Thats pretty much what I've done in the codebase I'm working on, it might be an idea to add it to react-select core but I guess it really depends on how many people would find that useful out of the box. At least for me, it makes sense for a basic single select to be able to accept either the react-select object or a string value - it feels more akin to how you'd expect a proper select field to work

**Comment by phoniks** [comment]
I'm only using react-select for the first time, but I've used Keystone in the past and admire Jed's work.  I think in this case, this behavior is counterintuitive.  At the very least I think it should be better documented in the Getting Started section since showing a value is a pretty basic function of a select component.  I'm at max capacity for open source contributions at the moment, but I can take a look at updating the docs on this after the new year.

**Comment by neilchaudhuri** [comment]
I am running into the same issue with v.2.1.2. At the moment I have as basic a setup as you can imagine:

```
<Select               options={colors.toArray()}
                            getOptionLabel={option => option}
                            getOptionValue={option => option}
                            value={this.props.color} />
```

In this snippet `colors` is a TypeScript `List<string>`; hence the `toArray()` call. 

So is the conclusion from this thread that I need to convert the `string[]` into `Option[]` with `label`/`value` shape for things to work? If so, that isn't the worst thing ever, but I agree with those who feel it is a bit cumbersome for such a basic scenario.

Just seeking clarification.

**Comment by neilchaudhuri** [comment]
Thanks. I created something that is probably similar. It may not get huge attendance at React Conf, but it's working for me.

**Comment by michaelmota** [comment]
I don't know if it's the most "optimal" way to do it, but none of the solutions worked for me so I "forced" it. Here is the approach I did.

`
returnValueMulti = (val) => {
        let arr =[]
        var item=''
        if (val !== null && val !== undefined) {
            for (let x in val) {
                item = this.state[index].find(item => item.value === val[x]);
                arr.push(item)
            }
            return arr
        }
    }
`

`
value={this.returnValueMulti(property.agents)}
`

**Comment by michaelmota** [comment]
I don't know if it's the most "optimal" way to do it, or "the right way" but here is what worked for me. I needed this for a "edit" view..
```
returnValueMulti = (index, val) => {
    let arr =[]
    var item=''
    if (val !== null && val !== undefined) {
        for (let x in val) {
            item = this.state[index].find(item => item.value === val[x]);
            arr.push(item)
        }
        return arr
    }
}
```
And the select:
```
<Select
    isMulti
    value={this.returnValueMulti("agents", property.agents)}
    options={agents}
/>
```

I'm 100% sure that there are better ways to do it. But this is the only thing that worked for me, just in case some one else is having the same issue assigning the value to the isMulti select.

**Comment by mattyx96** [comment]
the solution:

                   <Select
                        styles={selectStyles}
                        value={{label : this.state.string}}
                        onChange={(e) => this.handleChange(e)}
                        options={this.state.options}
                    />

**Comment by Flibo** [comment]
I feel like this is still an issue.

At the moment -- to my understanding -- if you want to make a controlled `Select` component you're required to feed it with `value={[{ value, label }]}` instead of `value={value}` like in `v1`. This seems completely unnecessary.

Having to include the label in `value` is particularly annoying with `AsyncSelect` where the options aren't directly available.

For now I'm going back to `v1`.

**Comment by powmedia** [comment]
I've just run in to the same problem after upgrading from v1.  I think the new behaviour is very counterintuitive, especially as this is not obvious in the docs or even the migration guide.

**Comment by FazelMahdi** [comment]
> Selected value does not show in the input field for a controlled react-select component with single select.
> 
> Version: 2.0.0-beta.6
> Demo: https://codesandbox.io/s/ypwq481j7z
> 
> ```js-jsx
> import React, { Component} from "react";
> import Select from "react-select";
> 
> type State = {
>   selectedValue: string
> };
> 
> export default class SingleSelect extends Component<*, State> {
>   state = {
>     selectedValue: ""
>   };
> 
>   optionChanged = value => {
>     console.log(value);
>     this.setState({ selectedValue: value});
>   };
>   render() {
>     console.log(this.state.selectedValue);
>     return (
>       <div>
>         <Select
>           options={["abcd", "def", "ghi"]}
>           value={this.state.selectedValue}
>           onChange={this.optionChanged}
>           getOptionLabel={option => option}
>           getOptionValue={option => option}
>         />
>       </div>
>     );
>   }
> }
> ```
> ```
I Solved this in an easy way:
you must you placeholder and update it with your selected option "label" like this:

 state = {
>     selectedValue: ' '
>     selectedValuePlaceholder: "select"
>   };

And;

  <Select
>           options={["abcd", "def", "ghi"]}
>           getOptionLabel={option => option}
>           getOptionValue={option => option}
>           placeholder={this.state.selectedValuePlaceholder}
>            onChange={({value, label}) => this.setState({selectedValue: value, selectedValuePlaceholder : 
>            label})}
> />

**Comment by askflow1111** [comment]
Hi,

you can also do this:

`
          value={{label: this.state.selectedValue, value: this.state.selectedValue} }
`

Thank you!

**Comment by willsmanley** [comment]
I think this behavior is very common among other dropdown/selection packages and makes sense a lot of sense. 

If a string is all you are selecting, then your value should be the string AND the label should be the string as askflow mentions.

Without this as the default behavior, the package wouldn't be very useful.

**Comment by h0jeZvgoxFepBQ2C** [comment]
Omg... just stumbled over this issue and I thought react-select was bugged... wtf.. This is really not intuitive 😕

**Comment by ktrzos** [comment]
@musmanrao1994 
You can't when you have other value in `value` and other in `label`.

```
[
  { value: 1, label: "User 1" },
  { value: 2, label: "User 2" },
  { value: 3, label: "User 3" },
]
```

So, with your solution, selected value (in `ValueContainer`) will show its number, and not a label.

**Comment by jackblackCH** [comment]
As well I am surprised that the value expects an object than just the value like any other dropdown.
This makes it hard as you you normally send and use the value only in the onChangeHandler, but to make the controlled dropdown work again, we have to play with save the whole object.

**Comment by evey201** [comment]
This was my solution to the problem: 
state = {
account_type: []
}

handleSelect = (account_type) => {
    this.setState({ 
      account_type: [{
      label: account_type.label,
      value: account_type.value}]
    });
    console.log(`Option selected:`, account_type.value);

  }

<Select 
           options={SelectOptions} 
           onChange={this.handleSelect} 
            value={this.state.account_type} 
 />



const SelectOptions = [
  { value: 'Wallet', label: 'Wallet' },
  { value: 'Bank', label: 'Bank' },
]

I hope you find this helpful in whatever way

**Comment by SalemFeyzo** [comment]
This is my solution
check out it on this url:
[https://codesandbox.io/s/prod-rgb-o5svh?file=/src/App.js:0-806](url) 
`
import axios from 'axios'
import {useEffect, useState} from 'react'
import Select from 'react-select'
import "./styles.css";

export default function App() {
  const [users, setUsers] = useState()
  const [userId, setUserId] = useState()

  useEffect(()=>{
    const getUsers = async () => {
      const {data} = await axios.get('https://jsonplaceholder.typicode.com/users')
      setUsers(data)
    }
    getUsers()
  },[])
const options = users && users.map(user =>{
  return {label: user.name, value: user.id}
})
console.log(userId)
  return (
    <div className="App">
     {users && (
       <Select 
       placeholder='Select user...'
       isSearchable
       value={options.label}
       options={options}
       onChange={(option) => setUserId(option.value) }
       />
     )}
    </div>
  )
}

`

**Comment by yashone7** [comment]
> @27percent if you need simple value, consider using a helper function:
> 
> ```js
> const valueFromId = (opts, id) => opts.find(o => o.value === id);
> 
> <Select value={valueFromId(this.props.options, this.props.id) />
> ```

Doing so might change the component from a controlled one to uncontrolled one if the id property is reset in the state

**Comment by tehpsalmist** [comment]
Still incredibly strange that the `getOptionLabel` function wouldn't apply to displaying the selected values, but it would seem that the surefire hack is to do `value={[value]}`. I suppose it could be worse, but I feel bad for the droves of people who will still have to waste their time googling for this solution until it's addressed more properly.

**Comment by hs2119** [comment]
The code from the documentation is not working for me. How is this possible??? Im doing this in electron if that affects anything but it really shouldn't
Neither is the solution where value={{value:"....", label: "..."}}

**Comment by Nierdhaza** [comment]
Hope my answer will help someone with this issue.
I did not find proof of my assumtions in code of this library.
But I assume that SingleValue component by default use label field to render from { label: 'label', value: 'value' } passed to it by default.
Here is an example of using react-select with value of structure {name, number, address} (you can use any structure you want)
You need to use your own SingleValue compoennt to render needed data structure 
[https://codesandbox.io/s/heuristic-orla-4m7blq?file=/src/App.js](url)

**Comment by Nierdhaza** [comment]
Hope my answer will help someone.
To use any data you want there should be created SingleValue custom component for Select component
example below:
[https://codesandbox.io/s/heuristic-orla-4m7blq?file=/src/App.js](url)

**Comment by mislavmiocevic** [comment]
I am using v5 and still this occurs (almost in year 2026 😆 ).

I solved it by adding opacity 1 to input.

Using chakra-react-select this looks like
```
const chakraStyles: ChakraStylesConfig<Option, false> = {
  input: styles => ({ ...styles, opacity: 1 })
}

<Select
chakraStyles={chakraStyles}
... // other props
/>
```

If you dont use Chakra wrapper, then probably extending components will solve it. Or any other CSS inject.

Full example
```
import { Select } from "chakra-react-select";
import { useState } from "react";

type Option = { value: string; label: string };

const options = [
    { value: "A", label: "A" },
    { value: "AA", label: "AA" }
];

const chakraStyles: ChakraStylesConfig<Option, false> = {
    input: styles => ({ ...styles, opacity: 1 })
}

export function MySelect() {
    const [searchText, setSearchText] = useState("");
    const [value, setValue] = useState<Option | undefined>();

    return (
       <Select<Option, false>
          chakraStyles={chakraStyles}
          inputValue={searchText}
          options={options}
          value={value}
          onChange={newValue => {
             setValue(newValue ?? undefined);
             setSearchText(newValue?.value ?? "");
          }}
          onInputChange={(newValue, { action }) => {
             if (action === "input-change") {
                setSearchText(newValue);
             }
          }}
       />
    );
}
```
